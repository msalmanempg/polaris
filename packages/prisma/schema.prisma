//******** AUTO GENERATED FILE, DO NOT EDIT.  *********

datasource db {
    provider = "postgresql"
    url      = env("DATABASE_URL")
}

generator client {
    provider = "prisma-client-js"
}

generator dbml {
    provider = "prisma-dbml-generator"
}


model Agreement {
  id                   Int                 @id @default(autoincrement())
  leadId               Int                 @map("lead_id")
  agreementNumber      String              @unique @map("agreement_number")
  publishedPrice       Int                 @map("published_price")
  bookingPrice         Int                 @map("booking_price")
  discountValue        Int?                @map("discount_value")
  posessionValue       Int                 @map("posession_value")
  payments             ExpectedPayment[]
  customers            AgreementCustomer[]
  salesPerson          SalePerson[]
  nominees             Nominee[]
  Unit                 Unit                @relation(fields: [unitId], references: [id])
  unitId               Int                 @map("unit_id")
  status               agreementStatus
  bookingDate          DateTime            @map("booking_date")
  notarisedDate        DateTime            @map("notarised_date")
  agreementDate        DateTime            @map("agreement_date")
  installmentStartDate DateTime            @map("installment_start_date")
  installmentEndDate   DateTime            @map("installment_end_date")
  createdAt            DateTime            @default(now()) @map("created_at")
  updatedAt            DateTime            @updatedAt @map("updated_at")

  @@map("agreements")
}

model DraftAgreement {
  id              Int      @id @default(autoincrement())
  Unit            Unit     @relation(fields: [unitId], references: [id])
  unitId          Int      @map("unit_id")
  agreementNumber String   @unique @map("agreement_number")
  data            Json     @map("data")
  createdAt       DateTime @default(now()) @map("created_at")
  updatedAt       DateTime @updatedAt @map("updated_at")

  @@map("draft_agreements")
}

model SalePerson {
  id          Int       @id @default(autoincrement())
  designation String
  fullName    String    @map("full_name")
  email       String
  govtId      String    @map("govt_id")
  phoneNumber String    @map("phone_number")
  sortLevel   Int       @map("sort_level")
  Agreement   Agreement @relation(fields: [agreementId], references: [id])
  agreementId Int       @map("agreement_id")
  createdAt   DateTime  @default(now()) @map("created_at")
  updatedAt   DateTime  @updatedAt @map("updated_at")

  @@map("sale_persons")
}

model AgreementCustomer {
  id                   Int                @id @default(autoincrement())
  type                 customerType       @map("customer_type")
  Customer             Customer           @relation(fields: [customerId], references: [id])
  customerId           Int                @map("customer_id")
  Agreement            Agreement          @relation(fields: [agreementId], references: [id])
  agreementId          Int                @map("agreement_id")
  isPrimary            Boolean            @default(true)
  meta                 Json?
  email                String
  primaryPhoneNumber   String             @map("primary_phone_number")
  secondaryPhoneNumber String?            @map("secondary_phone_number")
  address              String
  Companies            AgreementCompany[]
  createdAt            DateTime           @default(now()) @map("created_at")
  updatedAt            DateTime           @updatedAt @map("updated_at")

  @@map("agreement_customers")
}

model AgreementCompany {
  id                  Int               @id @default(autoincrement())
  AgreementCustomer   AgreementCustomer @relation(fields: [agreementCustomerId], references: [id])
  agreementCustomerId Int               @map("agreement_customer_id")
  name                String
  ntn                 String
  registrationNumber  String            @unique @map("registration_number")
  address             String
  city                String
  email               String
  contactNumber       String            @map("contact_number")
  meta                Json?
  createdAt           DateTime          @default(now()) @map("created_at")
  updatedAt           DateTime          @updatedAt @map("updated_at")

  @@map("agreement_companies")
}

model Nominee {
  id            Int        @id @default(autoincrement())
  govtId        String     @map("govt_id")
  govtIdType    govtIdType @map("govt_id_card")
  nomineeFor    String     @map("nominee_for")
  fullName      String     @map("full_name")
  contactNumber String     @map("contact_number")
  email         String
  relationship  String
  Agreement     Agreement  @relation(fields: [agreementId], references: [id])
  agreementId   Int        @map("agreement_id")
  meta          Json?
  createdAt     DateTime   @default(now()) @map("created_at")
  updatedAt     DateTime   @updatedAt @map("updated_at")

  @@map("nominees")
}

model AgreementAttachment {
  id           Int                   @id @default(autoincrement())
  agreementId  Int                   @map("agreement_id")
  Attachment   Attachment            @relation(fields: [attachmentId], references: [id])
  attachmentId Int
  type         agreementDocumentType
  createdAt    DateTime              @default(now()) @map("created_at")
  updatedAt    DateTime              @updatedAt @map("updated_at")

  @@map("agreement_attachments")
}

model NomineeAttachment {
  id           Int                   @id @default(autoincrement())
  nomineeId    Int                   @map("nominee_id")
  Attachment   Attachment            @relation(fields: [attachmentId], references: [id])
  attachmentId Int
  type         nomineeAttachmentType
  createdAt    DateTime              @default(now()) @map("created_at")
  updatedAt    DateTime              @updatedAt @map("updated_at")

  @@map("nominee_attachment")
}

enum govtIdType {
  cnic
  nicop
  poc
}

enum customerType {
  individual
  company
}

enum agreementStatus {
  draft
  review
  active
  transferred
  terminated
}

enum agreementDocumentType {
  agreement
  booking_form
  payment_plan
  floor_plan
  other
}

enum nomineeAttachmentType {
  id_front
  id_back
  profile_picture
  other
}


model Attachment {
  id                   Int                   @id @default(autoincrement())
  name                 String
  key                  String
  customerAttachments  CustomerAttachment[]
  nomineeAttachments   NomineeAttachment[]
  agreementAttachments AgreementAttachment[]
  createdAt            DateTime              @default(now()) @map("created_at")
  updatedAt            DateTime              @updatedAt @map("updated_at")

  @@map("attachments")
}


model Customer {
    id             Int                  @id @default(autoincrement())
    fullName       String               @map("full_name")
    govtId         String               @map("govt_id")
    govtIdType     govtIdType           @map("govt_id_type")
    relationType   relationType         @map("relation_type")
    guardianName   String               @map("guardian_name")
    nationality    String
    passportNumber String?              @map("passport_number")
    dateOfBirth    DateTime             @map("date_of_birth")
    gender         Gender
    meta           Json?
    attachments    CustomerAttachment[]
    agreements     AgreementCustomer[]
    createdAt      DateTime             @default(now()) @map("created_at")
    updatedAt      DateTime             @updatedAt @map("updated_at")

    @@unique([govtId, govtIdType], name: "Customer_govtId_govtIdType_unique_constraint")
    @@map(name: "customers")
}

model CustomerAttachment {
    id           Int        @id @default(autoincrement())
    Attachment   Attachment @relation(fields: [attachmentId], references: [id])
    attachmentId Int
    Customer     Customer   @relation(fields: [customerId], references: [id])
    customerId   Int
    createdAt    DateTime   @default(now()) @map("created_at")
    updatedAt    DateTime   @updatedAt @map("updated_at")

    @@map("customer_attachments")
}

enum relationType {
    father
    husband
}

enum Gender {
    male
    female
}

enum customerDocumentType {
    id_front
    id_back
    profile_picture
    other
}


model ExpectedPayment {
  id          Int                   @id @default(autoincrement())
  amount      Int
  percentage  Int?
  paymentDate DateTime              @map("payment_date")
  type        expectedPaymentType
  Agreement   Agreement             @relation(fields: [agreementId], references: [id])
  agreementId Int                   @map("agreement_id")
  meta        Json?
  dueDate     DateTime              @map("due_date")
  status      expectedPaymentStatus
  createdAt   DateTime              @default(now()) @map("created_at")
  updatedAt   DateTime              @updatedAt @map("updated_at")

  @@map("expected_payments")
}

enum expectedPaymentType {
  down_payment
  installment
  advance
  posession
}

enum expectedPaymentStatus {
  paid
  pending
}


model Project {
  id             Int                 @id @default(autoincrement())
  name           String
  address        String
  city           String
  province       String
  country        String
  longitude      Float
  latitude       Float
  email          String
  phone          String
  url            String
  completionDate DateTime?           @map("completion_date")
  logo           String
  meta           Json?
  bankDetails    ProjectBankDetail[]
  units          Unit[]
  createdAt      DateTime            @default(now()) @map("created_at")
  updatedAt      DateTime            @updatedAt @map("updated_at")

  @@map("projects")
}

model ProjectBankDetail {
  id            Int      @id @default(autoincrement())
  bankName      String   @map("bank_name")
  accountTitle  String   @map("account_title")
  accountNumber String   @map("account_number")
  iban          String
  swiftCode     String   @map("swift_code")
  branchAddress String   @map("branch_address")
  branchCode    String   @map("branch_code")
  meta          Json?
  Project       Project? @relation(fields: [projectId], references: [id])
  projectId     Int?
  createdAt     DateTime @default(now()) @map("created_at")
  updatedAt     DateTime @updatedAt @map("updated_at")

  @@map("project_bank_details")
}


model Unit {
  id             Int              @id @default(autoincrement())
  unitNumber     String           @unique @map("unit_number")
  location       String
  type           unitType
  bed            Int
  basePrice      Int              @map("base_price")
  publishedPrice Int              @map("published_price")
  netArea        Int              @map("net_area")
  grossArea      Int              @map("gross_area")
  meta           Json?
  status         unitStatus
  Project        Project          @relation(fields: [projectId], references: [id])
  projectId      Int              @map("project_id")
  agreements     Agreement[]
  drafts         DraftAgreement[]
  completionDate DateTime?        @map("completion_date")
  createdAt      DateTime         @default(now()) @map("created_at")
  updatedAt      DateTime         @updatedAt @map("updated_at")

  @@map(name: "units")
}

enum unitType {
  residential
  commercial
}

enum unitStatus {
  booked
  sold
  available
}
