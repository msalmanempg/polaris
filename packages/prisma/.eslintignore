**/__snapshots__/
**/.ng_build/
**/build/
**/coverage/
**/dist/
**/lib/
**/node_modules/
jest.config.js
*.d.ts