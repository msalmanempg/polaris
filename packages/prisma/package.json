{
  "name": "@polaris/prisma",
  "version": "1.0.0",
  "description": "data layer for Polaris",
  "main": "lib/index.js",
  "repository": "https://github.com/SectorLabs/polaris/tree/main/packages/prisma",
  "license": "MIT",
  "scripts": {
    "build": "yarn ttsc -p tsconfig.json",
    "lint:errors": "eslint --quiet --ext .js,.ts ./",
    "lint:types": "ttsc --noEmit"
  }
}
