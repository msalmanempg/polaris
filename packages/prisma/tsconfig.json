{
  "compilerOptions": {
    "allowJs": true,
    "module": "commonjs",
    "esModuleInterop": true,
    "moduleResolution": "node",
    "skipLibCheck": true,
    "isolatedModules": true,
    "forceConsistentCasingInFileNames": true,
    "lib": ["es2017", "dom"],
    "target": "es2017",
    "types": ["node"],
    "strict": true,
    "baseUrl": ".",
    "paths": {
      "prisma/*": ["."]
    },
    "outDir": "./lib/",
    "sourceMap": true,
    "inlineSources": true,
    "sourceRoot": "/",
    "plugins": [{ "transform": "@zerollup/ts-transform-paths" }]
  },
}
