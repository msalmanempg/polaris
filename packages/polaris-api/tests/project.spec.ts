import { prisma, Project, ProjectBankDetail } from "@polaris/prisma";
import { projectBankDetails, projects } from "./support/fixtures/projects";
import { testExpressApp } from "./support/requestFactory";

describe("projects", () => {
  let createdProjects: Project[];

  beforeEach(async () => {
    const projectsData = projects.map((project, index) => ({
      ...project,
      bankDetails: {
        create: [projectBankDetails[index]],
      },
    }));

    createdProjects = await Promise.all(
      projectsData.map((project) =>
        prisma.project.create({
          data: project as any,
        })
      )
    );
  });

  afterEach(async () => {
    await prisma.projectBankDetail.deleteMany({
      where: {
        projectId: {
          in: createdProjects.map((project) => project.id),
        },
      },
    });

    await prisma.project.deleteMany({
      where: {
        id: {
          in: createdProjects.map((project) => project.id),
        },
      },
    });
  });

  it("Should list all projects", async () => {
    const params = {
      pageOffset: 0,
      pageSize: 10,
    };
    const { body } = await testExpressApp().get("/api/project").query(params).expect(200);
    expect(body).toMatchSnapshot({
      count: expect.any(Number),
      pageOffset: expect.any(Number),
      pageSize: expect.any(Number),
      results: expect.arrayContaining([
        expect.objectContaining({
          id: expect.any(Number),
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
          completionDate: expect.any(String),
          bankDetails: expect.arrayContaining([
            expect.objectContaining({
              id: expect.any(Number),
              projectId: expect.any(Number),
            }),
          ]),
        }),
      ]),
    });
  });

  it("Should create a new project", async () => {
    const project = {
      name: "Project 4",
      address: "Cantt",
      city: "Karachi",
      province: "Sindh",
      country: "Pakistan",
      longitude: 4.14678,
      latitude: 2.36745,
      email: "project3@zameen.com",
      phone: "+923007651326",
      url: "https://www.zameendevelopments.com",
      logo:
        "http://gdj.graphicdesignjunction.com/wp-content/uploads/2012/07/business-logo-design-2.jpg",
      completionDate: "2022-02-15T00:03:58.000Z",
      bankDetails: [
        {
          bankName: "United Bank",
          accountTitle: "Project 3 Account",
          accountNumber: "22222222222222",
          iban: "PK067UB22222222222222",
          swiftCode: "0932",
          branchAddress: "mm alam",
          branchCode: "00722",
          createdAt: new Date("2021-02-19T05:20:07.269Z"),
          updatedAt: new Date("2021-02-19T05:20:07.269Z"),
        },
      ],
    };

    const { body } = await testExpressApp().post("/api/project/add").send(project).expect(201);

    expect(body).toMatchSnapshot({
      id: expect.any(Number),
      createdAt: expect.any(String),
      updatedAt: expect.any(String),
      completionDate: expect.any(String),
      bankDetails: expect.arrayContaining([
        expect.objectContaining({
          id: expect.any(Number),
          projectId: expect.any(Number),
        }),
      ]),
    });
  });

  it("Should update existing project", async () => {
    const existingProject = await prisma.project.findFirst({
      where: { name: "Project 3" },
      include: { bankDetails: true },
    });

    const projectId: number = existingProject ? existingProject.id : 0;
    const existingBankDetails: ProjectBankDetail[] = existingProject
      ? existingProject.bankDetails
      : [];

    const project = {
      name: "Project 3",
      address: "Address Updated",
      city: "Karachi",
      province: "Sindh",
      country: "Pakistan",
      longitude: 4.14678,
      latitude: 2.36745,
      email: "project3@zameen.com",
      phone: "+923007651326",
      url: "https://www.zameendevelopments.com",
      logo:
        "http://gdj.graphicdesignjunction.com/wp-content/uploads/2012/07/business-logo-design-2.jpg",
      completionDate: new Date("2021-02-19T05:20:07.269Z"),
      bankDetails: existingBankDetails,
    };

    await testExpressApp().put(`/api/project/update/${projectId}`).send(project).expect(204);
  });

  it("Should delete existing project", async () => {
    const existingProject = await prisma.project.findFirst({
      where: { name: "Project 2" },
    });

    const projectId: number = existingProject ? existingProject.id : 0;

    await testExpressApp().delete(`/api/project/delete/${projectId}`).expect(204);
  });

  it("Should get Project details with id", async () => {
    const existingProject = await prisma.project.findFirst({
      where: { name: "Project 2" },
    });
    const projectId: number = existingProject ? existingProject.id : 0;

    const { body } = await testExpressApp().get(`/api/project/detail/${projectId}`).expect(200);
    expect(body).toMatchSnapshot({
      id: expect.any(Number),
      createdAt: expect.any(String),
      updatedAt: expect.any(String),
      completionDate: expect.any(String),
      bankDetails: expect.arrayContaining([
        expect.objectContaining({
          id: expect.any(Number),
          projectId: expect.any(Number),
        }),
      ]),
    });
  });
});
