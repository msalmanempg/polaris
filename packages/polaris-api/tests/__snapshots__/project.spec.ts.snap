// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projects Should create a new project 1`] = `
Object {
  "address": "Cantt",
  "bankDetails": ArrayContaining [
    ObjectContaining {
      "id": Any<Number>,
      "projectId": Any<Number>,
    },
  ],
  "city": "Karachi",
  "completionDate": Any<String>,
  "country": "Pakistan",
  "createdAt": Any<String>,
  "email": "project3@zameen.com",
  "id": Any<Number>,
  "latitude": 2.36745,
  "logo": "http://gdj.graphicdesignjunction.com/wp-content/uploads/2012/07/business-logo-design-2.jpg",
  "longitude": 4.14678,
  "meta": null,
  "name": "Project 4",
  "phone": "+923007651326",
  "province": "Sindh",
  "updatedAt": Any<String>,
  "url": "https://www.zameendevelopments.com",
}
`;

exports[`projects Should get Project details with id 1`] = `
Object {
  "address": "Walton",
  "bankDetails": ArrayContaining [
    ObjectContaining {
      "id": Any<Number>,
      "projectId": Any<Number>,
    },
  ],
  "city": "Lahore",
  "completionDate": Any<String>,
  "country": "Pakistan",
  "createdAt": Any<String>,
  "email": "project2@zameen.com",
  "id": Any<Number>,
  "latitude": 1.36745,
  "logo": "http://gdj.graphicdesignjunction.com/wp-content/uploads/2012/07/business-logo-design-3.jpg",
  "longitude": 2.14678,
  "meta": null,
  "name": "Project 2",
  "phone": "+923007651325",
  "province": "Punjab",
  "updatedAt": Any<String>,
  "url": "https://www.zameendevelopments.com",
}
`;

exports[`projects Should list all projects 1`] = `
Object {
  "count": Any<Number>,
  "pageOffset": Any<Number>,
  "pageSize": Any<Number>,
  "results": ArrayContaining [
    ObjectContaining {
      "bankDetails": ArrayContaining [
        ObjectContaining {
          "id": Any<Number>,
          "projectId": Any<Number>,
        },
      ],
      "completionDate": Any<String>,
      "createdAt": Any<String>,
      "id": Any<Number>,
      "updatedAt": Any<String>,
    },
  ],
}
`;
